profile/encode.go

```diff
@@ -289,21 +296,35 @@ func (p *Profile) postDecode() error {
 	for _, s := range p.Sample {
 		labels := make(map[string][]string, len(s.labelX))
 		numLabels := make(map[string][]int64, len(s.labelX))
+		numUnits := make(map[string][]string, len(s.labelX))
+		unitEncountered := make(map[string]bool, len(s.labelX))
 		for _, l := range s.labelX {
 			var key, value string
 			key, err = getString(p.stringTable, &l.keyX, err)
 			if l.strX != 0 {
 				value, err = getString(p.stringTable, &l.strX, err)
 				labels[key] = append(labels[key], value)
 			} else if l.numX != 0 {
+				var unit string
+				if l.unitX != 0 {
+					unit, err = getString(p.stringTable, &l.unitX, err)
+					unitEncountered[key] = true
+				}
 				numLabels[key] = append(numLabels[key], l.numX)
+				numUnits[key] = append(numUnits[key], unit)
 			}
 		}
 		if len(labels) > 0 {
 			s.Label = labels
 		}
 		if len(numLabels) > 0 {
 			s.NumLabel = numLabels
+			for k := range numUnits {
+				if !unitEncountered[k] {
+					delete(numUnits, k)
```

Comment:
Maybe we can optimize a bit for the case of no tag units by just skipping append if the unit is an empty string and catching up for the desired length when we do need to append a non-empty unit.